default/python/universe_generation/specials.py

```diff
@@ -20,57 +71,87 @@ def distribute_specials(specials_freq, universe_objects):
                 fo.special_spawn_limit(sp) > 0 and fo.special_has_location(sp)]
     if not specials:
         return
+
     # dump a list of all specials meeting that conditions and their properties to the log
     print "Specials available for distribution at game start:"
     for special in specials:
-        print "...", special, ": spawn rate", fo.special_spawn_rate(special),\
-              "/ spawn limit", fo.special_spawn_limit(special)
-
-    # attempt to apply a special to each universe object in the list that has been passed to this function
-    # by finding a special that can be applied to it and hasn't been added too many times, and then attempt
-    # to add that special by testing its spawn rate
-    repeat_rate = {1: 0.08, 2: 0.05, 3: 0.01, 4: 0.00}
-    for univ_obj in universe_objects:
-        # for this universe object, find a suitable special
-        # start by shuffling our specials list, so each time the specials are considered in a new random order
-        random.shuffle(specials)
-        num_here = 0
-
-        # then, consider each special until one has been found or we run out of specials
-        # (the latter case means that no special is added to this universe object)
+        print("... {:30}: spawn rate {:2.3f} / spawn limit {}".
+              format(special, fo.special_spawn_rate(special), fo.special_spawn_limit(special)))
+
+    objects_needing_specials = [obj for obj in universe_objects if random.random() < base_chance]
+
+    track_num_placed = {obj: 0 for obj in universe_objects}
+
+    print("Base chance for specials is {}. Placing specials on {} of {} ({:1.4f})objects"
+          .format(base_chance, len(objects_needing_specials), len(universe_objects),
+                  float(len(objects_needing_specials)) / len(universe_objects)))
+
+    obj_tuple_needing_specials = set(zip(objects_needing_specials,
+                                         fo.objs_get_systems(objects_needing_specials),
+                                         calculate_number_of_specials_to_place(objects_needing_specials)))
+
+    # Equal to the largest distance in WithinStarlaneJumps conditions
+    # GALAXY_DECOUPLING_DISTANCE is used as follows.  For any two or more objects
+    # at least GALAXY_DECOUPLING_DISTANCE appart you only need to check
+    # fo.special_locations once and then you can place as many specials as possible,
+    # subject to number restrictions.
+    #
+    # Organize the objects into sets where all objects are spaced GALAXY_DECOUPLING_DISTANCE
+    # appart.  Place a special on each one.  Repeat until you run out of specials or objects.
+    GALAXY_DECOUPLING_DISTANCE = 6
```

Comment:
Ok, if I understand the algorithm correctly, what you're trying to achieve is to minimize the number of calls to the expensive `special_locations` API function by this organization of the passed in list of universe objects into sets where all objects are at least `GALAXY_DECOUPLING_DISTANCE` jumps apart, which allows you to precalculate and cache the location conditions.

That's nifty, but of course works only under the assumption that no `WithinStarlaneJumps` condition is used in the location condition of any special which specifies a jump distance limit that is greater than 6.

Though somewhat reasonable, this still isn't correct, as you simply can't make this assumption safely. Nevertheless, it's good enough for a stop-gap solution I guess, so I'm ok with that approach for now. However, we need to revisit this entire placement thing (and I'm talking about monsters, specials and natives) at some point to come up with a 100% correct solution.

Sorry, long term I won't relent on that. :wink: